File: scripts/mintlify/toc-to-docs-json.mjs
#!/usr/bin/env node
/**
 * Convert a DocFX toc.yml into Mintlify navigation (groups/pages).
 *
 * Usage:
 *   node toc-to-docs-json.mjs <package-dir> [--out <file.json>]
 *
 * Example:
 *   node toc-to-docs-json.mjs ../../docs/dpf/dpf-framework/versions/2027.R1.SP01 \
 *     --out ../../navigation/dpf-framework-2027-r1-sp01-groups.json
 */

import { readFileSync, writeFileSync, mkdirSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { parse as parseYaml } from "yaml";

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = resolve(__dirname, "../..");

const COLLAPSED_GROUPS = new Set([
  "Operator specifications",
  "averaging",
  "compression",
  "mapping",
  "mesh",
  "min_max",
  "result",
  "serialization",
  "utility",
]);

function parseArgs(argv) {
  const args = { packageDir: null, out: null };
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === "--out" && argv[i + 1]) {
      args.out = argv[++i];
      continue;
    }
    if (!argv[i].startsWith("-")) {
      args.packageDir = argv[i];
    }
  }
  return args;
}

/** Repo-relative Mintlify page path (no extension). */
function hrefToPage(href, contentPrefix) {
  const slug = href.replace(/\.mdx?$/i, "");
  return `${contentPrefix}/${slug}`.replace(/\/+/g, "/");
}

function shouldCollapse(name) {
  return COLLAPSED_GROUPS.has(name);
}

function convertItems(items, contentPrefix) {
  const pages = [];
  for (const item of items) {
    const converted = convertNode(item, contentPrefix);
    if (converted !== null) {
      pages.push(converted);
    }
  }
  return pages;
}

function convertNode(item, contentPrefix) {
  const { name, href, items } = item;

  if (items?.length) {
    const childPages = convertItems(items, contentPrefix);
    const group = {
      group: name,
      pages: childPages,
    };
    if (href) {
      group.root = hrefToPage(href, contentPrefix);
    }
    if (shouldCollapse(name)) {
      group.expanded = false;
    }
    return group;
  }

  if (href) {
    return hrefToPage(href, contentPrefix);
  }

  return null;
}

function tocToGroups(toc, contentPrefix) {
  return convertItems(toc, contentPrefix);
}

function packageDirToContentPrefix(packageDirAbs) {
  const rel = packageDirAbs
    .replace(/\\/g, "/")
    .replace(`${REPO_ROOT}/`.replace(/\\/g, "/"), "");
  return rel;
}

function main() {
  const { packageDir, out } = parseArgs(process.argv);
  if (!packageDir) {
    console.error(
      "Usage: node toc-to-docs-json.mjs <package-dir> [--out <groups.json>]"
    );
    process.exit(1);
  }

  const packageDirAbs = resolve(REPO_ROOT, packageDir);
  const tocPath = join(packageDirAbs, "toc.yml");
  const toc = parseYaml(readFileSync(tocPath, "utf8"));
  const contentPrefix = packageDirToContentPrefix(packageDirAbs);
  const groups = tocToGroups(toc, contentPrefix);

  const outPath = out
    ? resolve(REPO_ROOT, out)
    : join(packageDirAbs, "mintlify-groups.json");

  mkdirSync(dirname(outPath), { recursive: true });
  writeFileSync(outPath, `${JSON.stringify(groups, null, 2)}\n`, "utf8");

  const pageCount = JSON.stringify(groups).split('"group"').length - 1;
  console.log(`Wrote ${groups.length} top-level nav entries to ${outPath}`);
  console.log(`Content prefix: ${contentPrefix}`);
  console.log(`(includes ${pageCount}+ nested groups)`);
}

main();

File: scripts/mintlify/toc_to_docs_json.py
#!/usr/bin/env python3
"""
Convert DocFX toc.yml to Mintlify navigation groups JSON.

Usage:
  python toc_to_docs_json.py <package-dir> [--out <groups.json>]
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

try:
    import yaml
except ImportError:
    print("Install PyYAML: pip install pyyaml", file=sys.stderr)
    sys.exit(1)

REPO_ROOT = Path(__file__).resolve().parents[2]

COLLAPSED_GROUPS = {
    "Operator specifications",
    "averaging",
    "compression",
    "mapping",
    "mesh",
    "min_max",
    "result",
    "serialization",
    "utility",
}


def href_to_page(href: str, content_prefix: str) -> str:
    slug = href.removesuffix(".md").removesuffix(".mdx")
    return f"{content_prefix}/{slug}".replace("//", "/")


def should_collapse(name: str) -> bool:
    return name in COLLAPSED_GROUPS


def convert_node(item: dict, content_prefix: str):
    name = item.get("name", "")
    href = item.get("href")
    items = item.get("items")

    if items:
        group: dict = {"group": name, "pages": convert_items(items, content_prefix)}
        if href:
            group["root"] = href_to_page(href, content_prefix)
        if should_collapse(name):
            group["expanded"] = False
        return group

    if href:
        return href_to_page(href, content_prefix)

    return None


def convert_items(items: list, content_prefix: str) -> list:
    pages = []
    for item in items:
        node = convert_node(item, content_prefix)
        if node is not None:
            pages.append(node)
    return pages


def main() -> None:
    parser = argparse.ArgumentParser(description="Convert toc.yml to Mintlify groups JSON")
    parser.add_argument("package_dir", help="Path to package (contains toc.yml)")
    parser.add_argument("--out", help="Output JSON path (repo-relative or absolute)")
    args = parser.parse_args()

    package_dir = Path(args.package_dir)
    if not package_dir.is_absolute():
        package_dir = REPO_ROOT / package_dir

    toc_path = package_dir / "toc.yml"
    toc = yaml.safe_load(toc_path.read_text(encoding="utf-8"))

    content_prefix = package_dir.as_posix().replace(REPO_ROOT.as_posix() + "/", "")
    groups = convert_items(toc, content_prefix)

    if args.out:
        out_path = Path(args.out)
        if not out_path.is_absolute():
            out_path = REPO_ROOT / out_path
    else:
        out_path = package_dir / "mintlify-groups.json"

    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(groups, indent=2) + "\n", encoding="utf-8")

    main_path = REPO_ROOT / "navigation" / "main.json"
    main_path.write_text(
        json.dumps(
            {
                "products": [
                    {
                        "product": "DPF Framework",
                        "description": "Data Processing Framework — installation, user guide, and operator reference.",
                        "icon": "layers",
                        "versions": [
                            {
                                "version": "2027 R1 SP01",
                                "default": True,
                                "tag": "Pilot",
                                "groups": groups,
                            }
                        ],
                    }
                ]
            },
            indent=2,
        )
        + "\n",
        encoding="utf-8",
    )

    intro = groups[0] if groups else None
    print(f"Wrote {len(groups)} top-level nav entries to {out_path}")
    print(f"Wrote navigation/main.json (groups inlined, no $ref)")
    print(f"Content prefix: {content_prefix}")
    if isinstance(intro, str):
        print(f"Intro page path: /{intro}")


if __name__ == "__main__":
    main()
